File: MergeSort.java

public class MergeSort {

    // Method to divide the array
    public static void mergeSort(int[] array, int left, int right) {

        // Check if there is more than one element
        if (left < right) {

            // Find the middle index
            int middle = (left + right) / 2;

            // Sort the left half
            mergeSort(array, left, middle);

            // Sort the right half
            mergeSort(array, middle + 1, right);

            // Merge the two sorted halves
            merge(array, left, middle, right);
        }

    }

    // Method to merge two sorted subarrays
    public static void merge(int[] array, int left, int middle, int right) {

        // Calculate sizes of left and right subarrays
        int leftSize = middle - left + 1;
        int rightSize = right - middle;

        // Create temporary arrays
        int[] leftArray = new int[leftSize];
        int[] rightArray = new int[rightSize];

        // Copy data into left temporary array
        for (int i = 0; i < leftSize; i++) {
            leftArray[i] = array[left + i];
        }

        // Copy data into right temporary array
        for (int j = 0; j < rightSize; j++) {
            rightArray[j] = array[middle + 1 + j];
        }

        // Initial indexes
        int i = 0;
        int j = 0;
        int k = left;

        // Merge the temporary arrays
        while (i < leftSize && j < rightSize) {

            if (leftArray[i] <= rightArray[j]) {
                array[k] = leftArray[i];
                i++;
            } else {
                array[k] = rightArray[j];
                j++;
            }

            k++;
        }

        // Copy remaining elements of leftArray
        while (i < leftSize) {
            array[k] = leftArray[i];
            i++;
            k++;
        }

        // Copy remaining elements of rightArray
        while (j < rightSize) {
            array[k] = rightArray[j];
            j++;
            k++;
        }

    }

    // Method to print the array
    public static void printArray(int[] array) {

        for (int number : array) {
            System.out.print(number + " ");
        }

        System.out.println();
    }

    public static void main(String[] args) {

        // Unsorted array
        int[] numbers = {38, 27, 43, 3, 9, 82, 10};

        System.out.println("Before Sorting:");
        printArray(numbers);

        // Call Merge Sort
        mergeSort(numbers, 0, numbers.length - 1);

        System.out.println("After Sorting:");
        printArray(numbers);

    }
}

File: MergeSort.c
#include <stdio.h>

void merge(int arr[], int left, int mid, int right) {
    int i, j, k;

    int n1 = mid - left + 1;
    int n2 = right - mid;

    int leftArray[n1], rightArray[n2];

    // Copy data to temporary arrays
    for(i = 0; i < n1; i++) {
        leftArray[i] = arr[left + i];
    }

    for(j = 0; j < n2; j++) {
        rightArray[j] = arr[mid + 1 + j];
    }

    i = 0;
    j = 0;
    k = left;

    // Merge the temporary arrays
    while(i < n1 && j < n2) {

        if(leftArray[i] <= rightArray[j]) {
            arr[k] = leftArray[i];
            i++;
        }
        else {
            arr[k] = rightArray[j];
            j++;
        }

        k++;
    }

    // Copy remaining elements of leftArray
    while(i < n1) {
        arr[k] = leftArray[i];
        i++;
        k++;
    }

    // Copy remaining elements of rightArray
    while(j < n2) {
        arr[k] = rightArray[j];
        j++;
        k++;
    }
}

void mergeSort(int arr[], int left, int right) {

    if(left < right) {

        int mid = (left + right) / 2;

        // Sort left half
        mergeSort(arr, left, mid);

        // Sort right half
        mergeSort(arr, mid + 1, right);

        // Merge both halves
        merge(arr, left, mid, right);
    }
}

int main() {

    int arr[100];
    int n, i;

    printf("Enter the number of elements: ");
    scanf("%d", &n);

    printf("Enter %d elements:\n", n);

    for(i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    mergeSort(arr, 0, n - 1);

    printf("\nSorted Array (Ascending Order):\n");

    for(i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }

    printf("\n");

    return 0;
}
